toreview/KadActionsBuilder.java

```diff
@@ -0,0 +1,311 @@
+                    request.getOperationId(),
+                    DEFAULT_PARTS, DEFAULT_PARTS,
+                    KadAction.PayloadType.BOOLEAN,
+                    Boolean.toString(accepted)
+            );
+        else
+            return INVALID_ACTION;
+    }
+
+
+    /**
+     * @param actionID the unique ID of the action
+     * @param peer     the receiving peer
+     * @param node     the node to be stored
+     * @return if the given ID is valid: an action built to execute the store (node) command to the peer, otherwise returns an INVALID_ACTION
+     */
+    public KadAction buildStore(int actionID, SMSPeer peer, Node<BinarySet> node) {
+        if (isValidID(actionID))
+            return new KadAction(
+                    peer,
+                    KadAction.ActionType.STORE,
+                    actionID,
+                    DEFAULT_PARTS, DEFAULT_PARTS,
+                    KadAction.PayloadType.NODE_ID,
+                    node.getKey().toHex()
+            );
+        else
+            return INVALID_ACTION;
+    }
+
+    /**
+     * @param actionID the unique ID of the action
+     * @param peer     the receiving peer
+     * @param resource the resource to be stored
+     * @return if the given ID is valid: an action built to execute the store (resource) command to the peer, otherwise returns an INVALID_ACTION
+     */
+    public KadAction buildStore(int actionID, SMSPeer peer, StringResource resource) {
+        if (isValidID(actionID))
+            return new KadAction(
+                    peer,
+                    KadAction.ActionType.STORE,
+                    actionID,
+                    DEFAULT_PARTS, DEFAULT_PARTS,
+                    KadAction.PayloadType.RESOURCE,
+                    resource.getName() + RESOURCE_SEPARATOR + resource.getValue()
+            );
+        else
+            return INVALID_ACTION;
+    }
+
+    /**
+     * @param request the received store (node) request action
+     * @param peers   the peers of which forward addresses
+     * @return an array of responses (one for each peer in peers) for the received request, or an invalid KAD action if the given action wasn't coherent with this building method (actionType)
+     */
+    public KadAction[] buildStoreAnsw(KadAction request, SMSPeer[] peers) {
+        if (request.getActionType() == KadAction.ActionType.STORE) {
+            KadAction[] responses = new KadAction[peers.length];
+            for (int i = 0; i < peers.length; i++) {
+                responses[i] = new KadAction(
+                        request.getPeer(),
+                        KadAction.ActionType.STORE_ANSWER,
+                        request.getOperationId(),
+                        i + 1, peers.length,
+                        KadAction.PayloadType.PEER_ADDRESS,
+                        peers[i].getAddress()
+                );
+            }
+            return responses;
+        } else
+            return new KadAction[]{INVALID_ACTION};

```

Comment:
Questo metodo suppongo sia utilizzato solo dal manager di rete; perché non aggiungere semplicemente una specifica per dire che accetta solo request di tipo `STORE`? Forse è una mia fissazione, ma guardare l'array restituito per verificare che contenga un solo elemento e che questo sia una `KadAction` di tipo `INVALID_ACTION` nel caso di request invalida non è molto elegante. Piuttosto se la specifica prevede di gestire request che non sono `STORE`, lancia un'eccezione. Ma come già detto, molto presumibilmente questo metodo è interno, e pensato per essere utilizzato dal gestore di rete: dato che sei tu l'unico utilizzatore effettivo, puoi essere sicuro che il tipo di request è sempre STORE, e questo giustifica l'aggiunta della specifica di cui parlavo all'inizio.

Reply:
Mi sembrava si capisse in questa specifica, se non è **coerente** ritorna un'azione invalida.
`@return an array of responses (one for each peer in peers) for the received request, or an invalid KAD action if the given action wasn't coherent with this building method (actionType)`
Quindi se la request non è di tipo STORE, non è **coerente** con questo metodo.
Poi, il fatto di ritornare un array con un unico elemento che è l'azione invalida in effetti non è il massimo, però lanciare un'eccezione non mi pare sia adeguato, tutti gli altri metodi ritornano un'azione invalida, poi chi utilizza questa classe deve ricordarsi di quali metodi lanciano eccezioni e di quali metodi ritornano un'azione invalida? non mi sembra il massimo, a quel punto allora tanto vale che tutti lancino eccezioni. Ma è quello che si voleva evitare con l'azione invalida.

Reply:
Ma ci sono casi in cui si può verificare che `request` sia di tipo non `STORE?`

Reply:
Beh chi utilizza questo metodo può passare una qualsiasi azione, quindi può capitare.

Reply:
Scusa, ho commentato sopra prima di vedere la conversazione, riprendo qui: la decisione sta a te nel momento in cui determini le specifiche del metodo, certo puoi fare come hai fatto tu e verificare ogni singola volta che la richiesta sia della tipologia corretta, ma non ha più senso inserirla nelle precondizioni? Di fatto il nome stesso del metodo dà un'idea molto chiara di quale richieste gestisca, poi non posso dire che sia sbagliato, forse solo che non ne vale la pena (secondo me, eh)

Reply:
Boh, a me piace punire chi usa male gli strumenti che creo forse😂

Reply:
Comunque direi che in conclusione si tratta semplicemente di una tua buona abitudine, considerando che lo fai ovunque non si può certo dire che tu non sia coerente.